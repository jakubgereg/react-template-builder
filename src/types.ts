export type ComponentTemplate = {
  type?: string;
  className?: string;
  props?: Record<string, any>;
  children?: React.ReactNode | ComponentTemplate[];
};

export type HtmlComponent<T extends keyof React.JSX.IntrinsicElements> = Omit<
  ComponentTemplate,
  'component' | 'props'
> & {
  type: T;
  props?: Omit<React.JSX.IntrinsicElements[T], 'className'>;
};

export type CustomComponentMapping = Record<string, React.ElementType>;
