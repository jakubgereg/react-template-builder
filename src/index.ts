export * from './template-builder';
export * from './types';
