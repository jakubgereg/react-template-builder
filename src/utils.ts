import { HtmlElements, htmlElements } from './html-elements';

export const isHtmlElement = (component: string): component is HtmlElements =>
  (htmlElements as readonly string[]).includes(component);
