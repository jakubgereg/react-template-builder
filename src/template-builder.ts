import React from 'react';

import { ComponentTemplate, CustomComponentMapping } from './types';
import { defineComponent, isHtmlElement } from './utils';

const renderComponent =
  (mapping?: CustomComponentMapping) =>
  ({ type = 'div', className, children: childrenFromProps, props }: ComponentTemplate): React.ReactElement => {
    const children = Array.isArray(childrenFromProps)
      ? childrenFromProps.map((child, idx) =>
          React.createElement(React.Fragment, { key: idx }, renderComponent(mapping)(child))
        )
      : childrenFromProps;

    if (typeof type !== 'string') throw new Error('component key must be a string');

    const combinedProps = { ...props, className };
    if (isHtmlElement(type)) return React.createElement(type, combinedProps, children);
    if (mapping && mapping[type]) {
      const Component = mapping[type];
      return React.createElement(Component, combinedProps, children);
    }
    console.error(`Component type=[${type}] not found`);
    return React.createElement('div', combinedProps, children);
  };

export const createTemplateBuilder = (mapping?: CustomComponentMapping) => ({
  renderComponent: renderComponent(mapping),
  defineComponent
});
