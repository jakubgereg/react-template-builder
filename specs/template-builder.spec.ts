import ReactTestRendered from 'react-test-renderer';
import { createTemplateBuilder } from '../src/template-builder';
import { TestComponent, Box, List } from './TestComponent';

const basicBuilder = createTemplateBuilder();
const customBuilder = createTemplateBuilder({
  TestComponent,
  Box,
  List
});

describe('TemplateBuilder', () => {
  it('should match a snapshot', () => {
    const component = basicBuilder.renderComponent({ type: 'main', className: 'main', children: 'Hello, World!' });
    let result = ReactTestRendered.create(component).toJSON();
    expect(result).toMatchSnapshot();
  });

  it('should match a snapshot with custom components', () => {
    const component = customBuilder.renderComponent({
      type: 'TestComponent',
      className: 'test-class-name',
      props: { value: { array: ['one', 'two', 1, 2], object: { dynamic: true }, number: 1, string: 'X' } }
    });
    let result = ReactTestRendered.create(component).toJSON();
    expect(result).toMatchSnapshot();
  });

  it('should match a snapshot with custom nested components', () => {
    const component = customBuilder.renderComponent({
      type: 'Box',
      className: 'main-box',
      children: [
        {
          type: 'Box',
          className: 'nested-box',
          children: [
            {
              type: 'TestComponent',
              className: 'test-class-name',
              props: { value: { array: ['one', 'two', 1, 2], object: { dynamic: true }, number: 1, string: 'X' } }
            }
          ]
        },
        {
          type: 'List',
          className: 'list-class-name',
          props: { items: ['one', 'two', 'three'] }
        },
        {
          type: 'a',
          className: 'link-class-name',
          props: { href: 'github.com', children: 'Google' }
        }
      ]
    });

    let result = ReactTestRendered.create(component).toJSON();
    console.log(result);
    expect(result).toMatchSnapshot();
  });
});
