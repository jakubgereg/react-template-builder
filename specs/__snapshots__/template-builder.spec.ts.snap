// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TemplateBuilder should match a snapshot 1`] = `
<main
  className="main"
>
  Hello, World!
</main>
`;

exports[`TemplateBuilder should match a snapshot with custom components 1`] = `
<pre
  className="test-class-name"
>
  {
  "array": [
    "one",
    "two",
    1,
    2
  ],
  "object": {
    "dynamic": true
  },
  "number": 1,
  "string": "X"
}
</pre>
`;

exports[`TemplateBuilder should match a snapshot with custom nested components 1`] = `
<div
  className="main-box"
>
  <div
    className="nested-box"
  >
    <pre
      className="test-class-name"
    >
      {
  "array": [
    "one",
    "two",
    1,
    2
  ],
  "object": {
    "dynamic": true
  },
  "number": 1,
  "string": "X"
}
    </pre>
  </div>
  <ul
    className="list-class-name"
  >
    <li>
      one
    </li>
    <li>
      two
    </li>
    <li>
      three
    </li>
  </ul>
  <a
    className="link-class-name"
    href="github.com"
  />
</div>
`;
